vpr/src/route/rr_graph_reader.cpp

```diff
@@ -46,779 +34,754 @@
-    while (channelLists) {
-        index = get_attribute(channelLists, "index", loc_data).as_int(0);
-        chan_width.x_list[index] = get_attribute(channelLists, "info", loc_data).as_float();
-        channelLists = channelLists.next_sibling(channelLists.name());
-    }
-    channelLists = get_first_child(parent, "y_list", loc_data);
-    while (channelLists) {
-        index = get_attribute(channelLists, "index", loc_data).as_int(0);
-        chan_width.y_list[index] = get_attribute(channelLists, "info", loc_data).as_float();
-        channelLists = channelLists.next_sibling(channelLists.name());
-    }
-
-}
-
-/* Grid was initialized from the architecture file. This function checks
- * if it corresponds to the RR graph. Errors out if it doesn't correspond*/
-void verify_grid(pugi::xml_node parent, const pugiutil::loc_data & loc_data, const DeviceGrid& grid) {
-    pugi::xml_node grid_node;
-    int num_grid_node = count_children(parent, "grid_loc", loc_data);
-
-    grid_node = get_first_child(parent, "grid_loc", loc_data);
-    for (int i = 0; i < num_grid_node; i++) {
-        int x = get_attribute(grid_node, "x", loc_data).as_float();
-        int y = get_attribute(grid_node, "y", loc_data).as_float();
-
-        const t_grid_tile& grid_tile = grid[x][y];
-
-        if (grid_tile.type->index != get_attribute(grid_node, "block_type_id", loc_data).as_int(0)) {
-            vpr_throw(VPR_ERROR_OTHER, __FILE__, __LINE__,
-                    "Architecture file does not match RR graph's block_type_id at (%d, %d): arch used ID %d, RR graph used ID %d.", x, y,
-                     (grid_tile.type->index), get_attribute(grid_node, "block_type_id", loc_data).as_int(0));
-        }
-        if (grid_tile.width_offset != get_attribute(grid_node, "width_offset", loc_data).as_float(0)) {
-            vpr_throw(VPR_ERROR_OTHER, __FILE__, __LINE__,
-                    "Architecture file does not match RR graph's width_offset at (%d, %d)", x, y);
-        }
-
-        if (grid_tile.height_offset != get_attribute(grid_node, "height_offset", loc_data).as_float(0)) {
-            vpr_throw(VPR_ERROR_OTHER, __FILE__, __LINE__,
-                    "Architecture file does not match RR graph's height_offset at (%d, %d)", x, y);
-        }
-        grid_node = grid_node.next_sibling(grid_node.name());
-    }
-}
-
-static int pin_index_by_num(const t_class &class_inf, int num) {
-	for (int index = 0; index < class_inf.num_pins; ++index) {
-		if (num == class_inf.pinlist[index]) {
-			return index;
+/* Process node info and push it back to device_ctx.rr_nodes[]. When <loc> or <timing> arrives,
+ * the corresponding callback picks up the last item in device_ctx.rr_nodes[] and continues to fill it.
+ * current_node_id is kept because we need to build a map from node IDs to segment IDs and t_rr_node
+ * does not hold IDs.
+ * !! Also note that it's assumed that nodes appear in ID order and no ID gaps occur. */
```

Comment:
Why do they need to appear in order?

Reply:
It's because their count isn't known in advance, so allocating first and then doing `auto& node = device_ctx.rr_nodes[node.id]...` is no more possible without caching the node data. 

Nodes are read into the memory using `push_back`, so it's necessary that they appear in order so that the right ID points to the right node. In xc7a50t's rr_graph.xml, _not_ caching node data saves around 80MB.

Reply:
Why not something like;
```python

nodes = []
while len(size) < new_id:
  nodes.append(None)
assert nodes[new_id] == nullptr
nodes[new_id] = create_node()
```

Reply:
oops, makes sense. never thought of it

Reply:
I'm delaying implementing this for two reasons: 1) i'm afraid it can incur a performance penalty, 2) if we add a num_nodes attribute, it will be irrelevant.